baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/BuilderPrivateConstructor.java

```diff
@@ -0,0 +1,108 @@
+        linkType = BugPattern.LinkType.CUSTOM,
+        providesFix = BugPattern.ProvidesFix.REQUIRES_HUMAN_ATTENTION,
+        severity = BugPattern.SeverityLevel.WARNING,
+        summary = "A Builder with a static factory method on the encapsulating class must have a private constructor. "
+                + "Minimizing unnecessary public API prevents future API breaks from impacting consumers. ")
+public final class BuilderPrivateConstructor extends BugChecker implements BugChecker.ClassTreeMatcher {
+
+    @Override
+    public Description matchClass(ClassTree tree, VisitorState state) {
+        ClassSymbol enclosingClass = ASTHelpers.enclosingClass(ASTHelpers.getSymbol(tree));
+        if (enclosingClass == null) {
+            return Description.NO_MATCH;
+        }
+        if (!isValidBuilderClass(tree)) {
+            return Description.NO_MATCH;
+        }
+        List<MethodTree> constructors = ASTHelpers.getConstructors(tree);
+        if (constructors.size() != 1 || !ASTHelpers.isGeneratedConstructor(Iterables.getOnlyElement(constructors))) {
+            return Description.NO_MATCH;
+        }
+
+        if (!hasStaticBuilderFactory(enclosingClass, state)) {
+            // No factory method, the public constructor is used
+            return Description.NO_MATCH;
+        }
+
+        // If no fields exist, the constructor is placed after the curly brace
+        int constructorPosition = ((JCTree) tree).getStartPosition()
+                + state.getSourceForNode(tree).indexOf('{')
+                + 1;
+
+        for (Tree member : tree.getMembers()) {
+            if (member.getKind() == Kind.VARIABLE) {
+                int endPosition = state.getEndPosition(member);
+                if (endPosition > constructorPosition) {
+                    constructorPosition = endPosition;
+                }
+            }
+        }
+        return buildDescription(tree)
+                .addFix(SuggestedFix.builder()
+                        .replace(
+                                constructorPosition,
+                                constructorPosition,
+                                String.format("\nprivate %s() {}", tree.getSimpleName()))
+                        .build())
+                .build();
+    }
+
+    private static boolean isValidBuilderClass(ClassTree tree) {
+        return tree.getSimpleName().contentEquals("Builder")
+                && tree.getImplementsClause().isEmpty()
+                && tree.getExtendsClause() == null
+                && tree.getModifiers().getFlags().contains(Modifier.STATIC);
+    }
+
+    private static boolean hasStaticBuilderFactory(ClassSymbol classSymbol, VisitorState state) {
+        Set<MethodSymbol> matching = ASTHelpers.findMatchingMethods(
+                state.getName("builder"),
+                methodSymbol -> methodSymbol != null
+                        && methodSymbol.getReturnType().tsym.getSimpleName().contentEquals("Builder")
```

Comment:
might be fine to skip this type return validation? people can always tell us if there are false positives (and suppress as aa workaround :)

Reply:
Idea is that some (largely legacy) builders are meant to be used via their constructors, I don't want to create a bunch of false positives. This way we only flag builders that we expect aren't supposed to have exposed constructors.

Reply:
I've updated this to take the builder ClassTree into account given the endsWith matcher.